### PR TITLE
syntax: Merge parsing code for structures and variants

### PR DESCRIPTION
**Pull request by petrochenkov**
In particular, attributes are now parsed on fields of tuple variants

### CODE REVIEW

File: src/test/compile-fail/issue-12560-1.rs [hunk 1]
```diff
@@ -8,15 +8,13 @@
...
-// compile-flags: -Z parse-only
-
 // For style and consistency reasons, non-parametrized enum variants must
 // be used simply as `ident` instead of `ident ()`.
 // This test-case covers enum declaration.
 
 enum Foo {
-    Bar(), //~ ERROR nullary enum variants are written with no trailing `( )`
-    Baz(), //~ ERROR nullary enum variants are written with no trailing `( )`
+    Bar(), //~ ERROR empty tuple structs and enum variants are not allowed
```
Comment (nagisa) [inline src/test/compile-fail/issue-12560-1.rs]: I think this is pretty major regression to the error message. It was pretty clearly pointing out at the mistake before and telling how to fix it. Now it seems to suggest it is not allowed altogether, although only the syntax is incorrect.

Comment (petrochenkov) [inline src/test/compile-fail/issue-12560-1.rs]: I'll add a note then "note: Remove the trailing `()`". Good enough?

Comment (petrochenkov) [inline src/test/compile-fail/issue-12560-1.rs]: Tuple structs and enum variants with zero fields are indeed not allowed altogether, it's not just syntax.

Comment (nagisa) [inline src/test/compile-fail/issue-12560-1.rs]: I think a “help: remove trailing () if you meant to write empty enum variant” would be good enough.

Comment (petrochenkov) [inline src/test/compile-fail/issue-12560-1.rs]: Done.

**Comment by petrochenkov** [comment]
r? @nrc 
(@alexcrichton is already too busy)

File: src/test/compile-fail/issue-12560-1.rs (hunk 1 above)
Comment (nagisa) [inline src/test/compile-fail/issue-12560-1.rs]: Awesome!

File: src/libsyntax/feature_gate.rs [hunk 2]
```diff
@@ -867,6 +867,8 @@ impl<'a, 'v> Visitor<'v> for PostExpansionVisitor<'a> {
                 self.context.span_handler.span_err(span, "empty tuple structs and enum variants \
                                                           are not allowed, use unit structs and \
                                                           enum variants instead");
+                self.context.span_handler.span_help(span, "remove trailing `()` if you meant to \
+                                                           write empty enum variant or structure");
```
Comment (nrc) [inline src/libsyntax/feature_gate.rs]: The language here is a little bit confusing I would change to "remove trailing `()` to make a unit struct or unit enum variant" as the terminology then matches the error message.

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -4707,11 +4710,11 @@ impl<'a> Parser<'a> {
         Ok((class_name, ItemStruct(vdata, generics), None))
     }
 
-    pub fn parse_record_struct_body(&mut self) -> PResult<Vec<StructField>> {
+    pub fn parse_record_struct_body(&mut self, allow_pub: bool) -> PResult<Vec<StructField>> {
```
Comment (nrc) [inline src/libsyntax/parse/parser.rs]: please use an enum not a bool here

File: src/libsyntax/parse/parser.rs [hunk 4]
```diff
@@ -4725,9 +4728,7 @@ impl<'a> Parser<'a> {
         Ok(fields)
     }
 
-    pub fn parse_tuple_struct_body(&mut self,
-                                   generics: &mut ast::Generics)
-                                   -> PResult<Vec<StructField>> {
+    pub fn parse_tuple_struct_body(&mut self, allow_pub: bool) -> PResult<Vec<StructField>> {
```
Comment (nrc) [inline src/libsyntax/parse/parser.rs]: and here

**Comment by nrc** [comment]
r+ with the enum for bool and typos in the error message fixed.

**Comment by petrochenkov** [comment]
Updated.

**Comment by nrc** [comment]
Thanks!

@bors: r+

**Pull request merged**